sys/random: fix distribution of random_uint32_range()

PeterKietzmann opened the pull request:
### Contribution description
#8778 depicts that  [`random_uint32_range()`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/random.h#L95) does not return uniformly distributed numbers in the desired interval. This PR adds a fix that provides uniform distributions (as displayed below) but comes with a certain performance regression.

### Issues/PRs references
 #8778

### Results - Distribution
#### Native with `prng_xorshift`, [0,0xC0000000)-interval, 50k numbers

Old:
![old](https://user-images.githubusercontent.com/7765855/41860514-29cf366a-789f-11e8-8764-75e5f6fda509.png)

With fix:
![new](https://user-images.githubusercontent.com/7765855/41860532-30dd13b4-789f-11e8-8c10-7e0f8ae0538e.png)

### Results - Performance
#### samr21-xpro with `prng_xorshift`, [0,0xC0000000)-interval, 10seconds
Old: 244 KiB/s
With fix: 213 KiB/s

#### samr21-xpro with `prng_xorshift`, [0,0x80000001)-interval, 10seconds
Old: 244 KiB/s
With fix: 206 KiB/s

#### samr21-xpro with `prng_xorshift`, [0,0x80000000)-interval, 10seconds
Old: 244 KiB/s
With fix: 217 KiB/s

#### samr21-xpro with `prng_xorshift`, [0x18000)-interval, 10seconds
Old: 244 KiB/s
With fix: ~~168 KiB/s~~ 202 KiB/s

#### samr21-xpro with `prng_xorshift`, [0x10000)-interval, 10seconds
Old: 244 KiB/s
With fix: ~~186 KiB/s~~ 217 KiB/s


#### samr21-xpro with `prng_tinymt32`, [0,0xC0000000)-interval, 10seconds
Old: 229 KiB/s
With fix: 198 KiB/s

#### samr21-xpro with `prng_tinymt32`, [0,0x18000)-interval, 10seconds
Old: 229 KiB/s
With fix: ~~151 KiB/s~~ 180 KiB/s

#### samr21-xpro with `prng_tinymt32`, [0,0x10000)-interval, 10seconds
Old: 229 KiB/s
With fix: ~~177 KiB/s~~ 205 KiB/s

basilfx commented on sys/random/random.c:
```diff
@@ -54,3 +54,28 @@ void random_bytes(uint8_t *target, size_t n)
+            if (highest_pow2 << 1) {
+                random = (random % (highest_pow2 << 1) );
+            }
+            /* get new value in same interval if max. already reached */
```
Is there a line missing, or should this comment move to line 71?

PeterKietzmann replied:
leftover

basilfx commented on tests/rng/main.c:
```diff
@@ -90,12 +90,26 @@ static int cmd_dump(int argc, char **argv)
+        test_dump_range(samples, low_thresh, high_thresh);
+    }
+    else {
+        printf("usage: %s [samples] [lower-bound upper-bound]", argv[0]);
```
Missing newline.

basilfx commented on tests/rng/main.c:
```diff
@@ -220,12 +234,25 @@ static int cmd_speed(int argc, char **argv)
+        test_speed_range(duration, low_thresh, high_thresh);
+    }
+    else {
+        printf("usage: %s [duration] [lower-bound upper-bound]", argv[0]);
```
Missing newline.

basilfx commented:
My comments have been addressed. I have tested this, and looks good.

cgundogan commented on sys/random/random.c:
```diff
@@ -54,3 +54,27 @@ void random_bytes(uint8_t *target, size_t n)
+    uint32_t random = random_uint32();
+
+    /* check if range is a power of two */
+    if (!(range % highest_pow2)) {
```
the compiler won't be able to optimize the modulo operation. To check whether `range` is a power of two, you could use
```c
range && !( range & ( range - 1))
```

cgundogan commented on sys/random/random.c:
```diff
@@ -54,3 +54,27 @@ void random_bytes(uint8_t *target, size_t n)
+            random = random_uint32();
+            /* get new value from next power of two interval
+             * else, get new random number */
+            if (highest_pow2 << 1) {
```
no need to calculate `highest_pow2 << 1` with each loop cycle. You could shift `highest_pow2` once, before entering the loop.

cgundogan commented on sys/random/random.c:
```diff
@@ -54,3 +54,27 @@ void random_bytes(uint8_t *target, size_t n)
+            /* get new value from next power of two interval
+             * else, get new random number */
+            if (highest_pow2 << 1) {
+                random = (random % (highest_pow2 << 1) );
```
since `highest_pow2 << 1` is again a power of two, you can use binary `and` instead of an expensive modulo operation

tcschmidt replied:
... and then you can move the if clause out of the loop, by simply setting highest_pow2 to zero 
if (!(highest_pow2 << 1))

PeterKietzmann commented:
@cgundogan, @tcschmidt thanks for your reviews. Please check my latest commit. Results seem ok at first sight: distributions look uniform and with xorshift, [0x18000)-interval on a samr21-xpro we get 202 KiB/s (before: 169KiB/s). Will report some further results tomorrow.

gebart commented:
The change looks OK. What is the worst case performance like? 
range = 0x80000001 should be particularly bad, but any power of two + a small number should be bad. We should see approximately 50 % of the throughput compared to the performance of a power of two range.

PeterKietzmann commented:
> range = 0x80000001 should be particularly bad

I've added the numbers above. Basically this is 206 vs 244 KiB/s.  @gebart why do you expect 50%?

gebart commented:
@PeterKietzmann it was a wrong assumption. But what is happening at 0x80000001 is that approximately half of the generated numbers will be discarded

PeterKietzmann commented:
Added the assert. @smlng, @cgundogan please re-check the code and remove your change requests (if addressed)

PeterKietzmann commented on sys/random/random.c:
```diff
@@ -54,3 +54,36 @@ void random_bytes(uint8_t *target, size_t n)
+    uint32_t random = random_uint32();
+
+    /* check if range is a power of two */
+    if (range && !(range & (range - 1))) {
```
> [edit2]: but if range is pow2 wouldn't that mean range == highest_pow2 then? If so it might be even more efficient to compare those instead of the bit operation?!

Tried that, but results are same (well, "my" solution produced even two more numbers in 20 seconds, but I guess this is below any optimization influence anyway)

smlng replied:
yeah I think nowadays its hard to outsmart the compiler 😉 So best we can do is go for well documented and structured code.

smlng commented on sys/random/random.c:
```diff
@@ -54,3 +54,36 @@ void random_bytes(uint8_t *target, size_t n)
+    uint32_t random = random_uint32();
+
+    /* check if range is a power of two */
+    if (!(range & (range - 1))) {
```
I will not block this, but IMHO the code could gain readability by improving its structure. What I mean is that it essentially comes to:

1. set a proper `highest_pow2` 
2. generate a random number (fitting into range), and 
3. return it `+ a`. 

For instance something like:

```
uint32_t random_uint32_range(uint32_t a, uint32_t b)
{
    assert(a < b);

    uint32_t highest_pow2, range = b - a;
    uint8_t range_msb = bitarithm_msb(range);

    /* check if range is a power of two */
    if (!(range & (range - 1))) {
        highest_pow2 = (1 << range_msb) - 1;
    }
    else if (range_msb < 31) {
        /* leftshift for next power of two interval */
        highest_pow2 = (1 << (range_msb + 1)) -1;
    }
    else {
        /* disable modulo operation in loop below */
        highest_pow2 = UINT32_MAX;
    }
    /* get random numbers until value is smaller than range */
    do {
        random = (random_uint32() & highest_pow2);
    } while (random >= range);
    /* return random in range [a,b] */
    return (random + a);
}
```

smlng replied:
hope I didn't changed the logic ...

smlng replied:
oops forgot to define random, add change:

```
uint32_t highest_pow2, random, range = b - a;
```

PeterKietzmann commented:
@smlng even though it's just a matter of taste, I've adopted your proposal regarding code structuring to have you on my side. Distribution as well as performance tests  didn't (really) change. Can I squash?

smlng commented:
> to have you on my side

no need for that, but appreciated 😄 please squash

PeterKietzmann commented on sys/random/random.c:
```diff
@@ -54,3 +54,30 @@ void random_bytes(uint8_t *target, size_t n)
+
+    /* check if range is a power of two */
+    if (!(range & (range - 1))) {
+        highest_pow2 = (1 << range_msb) - 1;
```
Will check the effect but if not noticeable I tend to leave as is.

PeterKietzmann commented on sys/random/random.c:
```diff
@@ -54,3 +54,30 @@ void random_bytes(uint8_t *target, size_t n)
+    if (!(range & (range - 1))) {
+        highest_pow2 = (1 << range_msb) - 1;
+    }
+    else if (range_msb < 31) {
```
Tried that before, without effect. So I decided for keeping it this way which seems more self-explaining to me.

PeterKietzmann commented on sys/random/random.c:
```diff
@@ -54,3 +54,30 @@ void random_bytes(uint8_t *target, size_t n)
+{
+    assert(a < b);
+
+    uint32_t highest_pow2, random, range = b - a;
```
Will change

PeterKietzmann commented on tests/rng/test.c:
```diff
@@ -378,3 +408,27 @@ void test_speed(uint32_t duration)
+    /* collect samples as long as timer has not expired */
+    timeout = xtimer_now_usec64() + (duration * US_PER_SEC);
+
+    while (xtimer_now_usec64() < timeout) {
```
Agree. Maybe start/stop via msg or callback in future? IMO this is also out of scope for this PR.

gebart replied:
Yes, xtimer_set is a good proposal. We can do it in a different PR

PeterKietzmann commented:
@gebart I've updated variable names and squashed immediately. Did not update the bit-logic stuff as I couldn't measure any difference. Further improvement of the test app should be done in a separate PR.
